universe/supplycommit/multi_sm_manager.go

```diff
@@ -0,0 +1,394 @@
+	Wallet Wallet
+
+	// Chain is our access to the current main chain.
+	//
+	// TODO(roasbeef): can make a slimmer version of
+	Chain tapgarden.ChainBridge
+
+	// StateLog is the main state log that is used to track the state of the
+	// state machine. This is used to persist the state of the state machine
+	// across restarts.
+	StateLog StateMachineStore
+}
+
+// MultiStateMachineManager is a manager for multiple supply commitment state
+// machines, one for each asset group. It is responsible for starting and
+// stopping the state machines, as well as forwarding sending events to them.
+type MultiStateMachineManager struct {
+	// cfg is the configuration for the multi state machine manager.
+	cfg MultiStateMachineManagerCfg
+
+	// smCache is a cache that maps asset group public keys to their
+	// supply commitment state machines.
+	smCache StateMachineCache
+
+	// ContextGuard provides a wait group and main quit channel that can be
+	// used to create guarded contexts.
+	*fn.ContextGuard
+
+	startOnce sync.Once
+	stopOnce  sync.Once
+}
+
+// NewMultiStateMachineManager creates a new multi state machine manager.
+func NewMultiStateMachineManager(
+	cfg MultiStateMachineManagerCfg) *MultiStateMachineManager {
+
+	return &MultiStateMachineManager{
+		cfg: cfg,
+		ContextGuard: &fn.ContextGuard{
+			DefaultTimeout: DefaultTimeout,
+			Quit:           make(chan struct{}),
+		},
+	}
+}
+
+// Start starts the multi state machine manager.
+func (m *MultiStateMachineManager) Start() {
+	m.startOnce.Do(func() {
+		// Initialize the state machine cache.
+		m.smCache = NewStateMachineCache()
+	})
+}
+
+// Stop stops the multi state machine manager, which in turn stops all asset
+// group key specific supply commitment state machines.
+func (m *MultiStateMachineManager) Stop() {
+	m.stopOnce.Do(func() {
+		// Cancel the state machine context to signal all state machines
+		// to stop.
+		close(m.Quit)
+
+		// Stop all state machines.
+		m.smCache.StopAll()
+	})
+}
+
+// fetchStateMachine retrieves a state machine from the cache or creates a
+// new one if it doesn't exist. If a new state machine is created, it is also
+// started.
+func (m *MultiStateMachineManager) fetchStateMachine(
+	assetSpec asset.Specifier) (*StateMachine, error) {
+
+	groupKey, err := assetSpec.UnwrapGroupKeyOrErr()
+	if err != nil {
+		return nil, fmt.Errorf("asset specifier missing group key: %w",
+			err)
+	}
+
+	// Check if the state machine for the asset group already exists in the
+	// cache.
+	sm, ok := m.smCache.Get(*groupKey)
+	if ok {
+		return sm, nil
+	}
+
+	// If the state machine is not found, create a new one.
+	env := &Environment{
+		AssetSpec:        assetSpec,
+		TreeView:         m.cfg.TreeView,
+		Commitments:      m.cfg.Commitments,
+		Wallet:           m.cfg.Wallet,
+		Chain:            m.cfg.Chain,
+		StateLog:         m.cfg.StateLog,
+		CommitConfTarget: DefaultCommitConfTarget,
+	}
+
+	// TODO(ffranr): Get initial state from disk here.
```

Comment:
IIRC I gave a commit to do this in the original PR before we started to break it up into sub-PRs.

Reply:
Found it!

Here's the commit: https://github.com/lightninglabs/taproot-assets/commit/d033b3c8d6da5c175616aaeb696a6746d52f7d01